E2E: add temporary spec to test the WordPress.com Pro upgrade button.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR is the first in the series to lay the groundwork for implementing e2e coverage for the overhauled Plans feature.
This PR also includes a temporary stub spec meant to provide a baseline level of coverage for the Pro plan upgrade button while the fuller spec is being worked on.

Key changes:
1. removal of E2E bypass for the overhauled plans.
2. addition of new methods in the Plans POM to support interaction with the new plans grid.
3. addition of new e2e spec mirroring the addition of Plans Upgrade into the checkout cart.

Details:
1. currently all e2e runs from TeamCity are being redirected to show the legacy plans grid due to the bypass. In order to test the new plans grid, the bypass must be removed.
2. it has been decided to keep the Plans POM in one piece for the time being, therefore requiring new parameters, methods to support interactions with the new Plans grid and restructuring of some types.
3. the new e2e spec is essentially a mirror copy of the existing Plans: Upgrade spec which doesn't actually perform an upgrade; it clicks on a button in the Plans page to upgrade, validates the cart, then exits the checkout page. This spec will serve as a stand-in for now to provide minimal coverage for the time being while I work to spec out the critical code paths.

#### Testing instructions

Ensure the following:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E
  - [x] i18n E2E


Related to #62492

worldomonation committed:
Remove conditional hiding the new plans from e2e users.
- verified on local webapp that this only affects sites that do not have a legacy plan

Rename existing Plans specs with legacy in the name and suite title.

worldomonation committed:
Add initial version of Plans: Add Upgrade to Cart spec.

Add ne required param to all existing invocatios of PlansPage.

worldomonation committed:
Update `validateActiveNavigationTab` method to check for the `.is-active` selector.

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -97,31 +152,31 @@ export class PlansPage {
+	 * @param {LegacyPlans} param0.plan Name of the plan (e.g. "Premium")
 	 * @param {PlanActionButton} param0.buttonText Expected action button text (e.g. "Upgrade")
 	 */
 	async clickPlanActionButton( {
```
I chose not to update the method name here:
- it adds a lot of unnecessary changes to the PR;
- if a developer tries to use this for the Pro plan, the lack of such option quickly becomes apparent.

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -17,31 +22,41 @@ const selectors = {
 	 * @param {Page} page Instance of the Playwright page
 	 */
-	constructor( page: Page ) {
+	constructor( page: Page, version: PlansGridVersion ) {
```
I looked at keying off something else here, but the two-week ago me did not find anything useful to key off of.

The problem is partially due to the slower loading times for the Legacy plans grid which, frustratingly, is **not caught** by the `waitForLoadState('networkidle')`. 

Given the above, I tried to implement a `try/catch` to first look for the new plan, and fall back to the Legacy plan, but that introduces new problems:
- adds execution time if the page being shown is the legacy plans grid.
- what if the overhauled plans page takes just a bit longer than expected to load?

With all things considered, I felt it was important to require, explicitly, that the test writer specify what version of the Plans page they are intending to work with.

dpasque replied:
>With all things considered, I felt it was important to require, explicitly, that the test writer specify what version of the Plans page they are intending to work with.

Given our most recent discussion on the editor, I think this is a great approach! It's very deterministic -- we should know which page we expect. And if we don't get the one we expect, we should probably fail! So I think this will serve us very well 😄

dpasque commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -17,31 +22,41 @@ const selectors = {
+	legacyPlansGrid: '.plans-features-main',
+
+	// Overhauled plans view
+	overhauledPlansGrid: 'div.plans > table',
```
Did you need this selector? Or is it for future use? I didn't see it used anywhere, so wanted to double check

worldomonation replied:
Yep, this is an orphaned selector!

dpasque commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -17,31 +22,41 @@ const selectors = {
+
+	// Overhauled plans view
+	overhauledPlansGrid: 'div.plans > table',
+	upgradeToProButton: 'th button.is-primary',
```
Can we not use button text due to potential i18n usage? Otherwise, I might opt for keying off the button text here, but that's a 100% optional suggestion! 👍

worldomonation replied:
Right, I chose not to key off the button text for i18n concerns.

For English, there are at least two variants:
- `Start with Pro` (en-US)
- `Try Pro risk-free` (en-GB)

The user account (`simpleSiteFreePlanUser`) uses the en-GB language and I figured it would be too jarring for the test writer (whom is likely using en-US) to see an unexpectedly different string, so I chose to go with the CSS selector.

dpasque commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -51,22 +66,62 @@ export class PlansPage {
+	 *
+	 * @param {PlansComparisonAction} action Action to perform on the Plans comparison button.
+	 */
+	async clickPlanComparisonActionButton( action: PlansComparisonAction ): Promise< void > {
```
100% optional suggestion: for here, what do you think of using that pattern you came up with where you have two methods which each guarantee their ending state? Like the [two methods here](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts#L161)?

I've just been a huge fan of that design ever since you proposed it! 😄

dpasque reviewed (approved):
You did such a nice job weaving in the new coverage in an effective but non-disruptive way! 🎉 

I added a few optional comments - use them if you find the helpful! Otherwise, LGTM 🚀

worldomonation committed:
Address PR comments:
- change the `clickPlanComparisonActionButton` method to an end-state guarantee type of method.
- remove dead selector.

worldomonation committed:
Changes:
- change expected text of navtabs in plans page to 'Plans'

worldomonation committed:
Changes:
- remove dead type from PlansPageTab
- move selector into legacy plans view section

worldomonation committed:
Placate the linter
